pdns/misc.cc

```diff
@@ -1560,3 +1561,39 @@ DNSName reverseNameFromIP(const ComboAddress& ip)
 
   throw std::runtime_error("Calling reverseNameFromIP() for an address which is neither an IPv4 nor an IPv6");
 }
+
+static size_t getMaxHostNameSize()
+{
+#if defined(HOST_NAME_MAX)
+  return HOST_NAME_MAX;
+#endif
+
+#if defined(_SC_HOST_NAME_MAX)
+  auto tmp = sysconf(_SC_HOST_NAME_MAX);
+  if (tmp != -1) {
+    return tmp;
+  }
+#endif
+
+  /* _POSIX_HOST_NAME_MAX */
+  return 255;
+}
+
+std::string getCarbonHostName()
+{
+  std::string hostname;
+  hostname.resize(getMaxHostNameSize() + 1, 0);
+
+  if (gethostname(const_cast<char*>(hostname.c_str()), hostname.size()) != 0) {
+    throw std::runtime_error(stringerror());
+  }
+
+  auto pos = hostname.find(".");
+  if (pos != std::string::npos) {
+    hostname.resize(pos);
+  }
+
+  boost::replace_all(hostname, ".", "_");
```

Comment:
I might be missing something, but if you just truncated the string until the first dot there are no dots left in the string. So to me this `replace_all` reads as a no-op.

Reply:
It was present in the original code so I left it there since I don't want to break anything else but it doesn't make sense to me either. I looked at the history but as far as I can tell it was never needed. I would suggest removing it on master but keeping it around on the already released branches.

Reply:
#2726 #2381

Reply:
So should we truncate or replace? We can't have both. Or was the intend to strip a trailing dot?

Reply:
I gather from Habbie's links that the original intent was to keep only the local part of the name, then later we tried to use the full host name by replacing dots with underscores but failed to do so. So I propose to stop truncating on master, do you agree?

Reply:
yes

Reply:
#2726 says 'we should replace dots with underscores in `carbon-ourname`'. #2381 says "no don't do that".

Ignoring the potential for breakage, I'd say the right way is to do underscores-for-dots for the machine hostname, and do nothing for `carbon-ourname`.

Not ignoring the potential for breakage, I'm not sure what we should do.

@pieterlexis ?

Reply:
Oh, yeah, I meant to propose replacing dots with hyphens instead of underscores, I don't think underscores would ever work. I agree `carbon-ourname` should stay unaltered.

Reply:
I don't see why hyphens would be better than underscores, but @pieterlexis would know better

Reply:
hyphens or underscores don't matter in carbon, only dots do. This function should indeed replace the dots with _something_ that doesn't confuse graphite. When a user sets dot in `carbon-hostname`, they are on their own. We try to provide a safe default for those who don't set `carbon-hostname`.

Reply:
Alright, so I'll change this code to replace dots with underscores, and only when `carbon-ourname` is not set.